grpcproxy/grpcproxy.go

```diff
@@ -0,0 +1,106 @@
+package grpcproxy
+
+import (
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+
+	"fmt"
+	"plugin"
+	"strings"
+
+	"github.com/akutz/gpds/lib"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+func LoadGRPCProxyPlugin(path string, targetURL string, gRPCProxyMux *runtime.ServeMux) error {
+	// Load our plugin
+	p, err := plugin.Open(path)
+	if err != nil {
+		return err
+	}
+
+	// lookup the plug-in's Types symbol; it's the type map used to
+	// register the plug-in's modules
+	tmapObj, err := p.Lookup("Types")
+	if err != nil {
+		return fmt.Errorf("error: failed to lookup type map: %v", err)
+	}
+
+	// assert that the Types symbol is a *map[string]func() interface{}
+	tmapPtr, tmapOk := tmapObj.(*map[string]func() interface{})
+	if !tmapOk {
+		return fmt.Errorf("error: invalid type map: %T", tmapObj)
+	}
+
+	// assert that the type map pointer is not nil
+	if tmapPtr == nil {
+		return fmt.Errorf("error: nil type map: type=%[1]T val=%[1]v", tmapPtr)
+	}
+
+	// dereference the type map pointer
+	tmap := *tmapPtr
+
+	// register the plug-in's modules
+	for k, v := range tmap {
+		lib.RegisterModule(k, v)
+	}
+
+	// Instantiate a copy of the module registered by the plug-in.
+	modGo := lib.NewModule("register")
+
+	ctx := context.Background()
+
+	endpoint := strings.Replace(targetURL, "http://", "", 1)
+	endpoint = strings.Replace(endpoint, "https://", "", 1)
+
+	// Create a new v2 config
+	config := &v2Config{
+		ctx:        ctx,
+		mux:        gRPCProxyMux,
+		entrypoint: endpoint,
+	}
+
+	// Initialize mod_go with a v2 config implementation
+	if err := modGo.Init(ctx, config); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type v2Config struct {
+	ctx        context.Context
+	mux        *runtime.ServeMux
+	entrypoint string
+	opts       []grpc.DialOption
+}
+
+// Get returns the value for the specified key
+func (c *v2Config) Get(ctx context.Context, key string) interface{} {
```

Comment:
Have you come up with the `Get/Set` interface? If so, why not make it type-safe, like:

```
func (c *v2Config) Ctx() *context.Context { return c.ctx }
```

(or just exporting the fields, which is even easier)

Reply:
See the last comment, this mechanism solves a whole bunch of type conversion headaches. I'm not confident enough to redesign it, so if you want to take a shot, be my guest.